src/aws_encryption_sdk/streaming_client.py

```diff
@@ -83,28 +80,27 @@ class _ClientConfig(object):
     """
 
     source = attr.ib(hash=True, converter=aws_encryption_sdk.internal.utils.prep_stream_data)
-    materials_manager = attr.ib(
-        hash=True, default=None, validator=attr.validators.optional(attr.validators.instance_of(CryptoMaterialsManager))
-    )
-    key_provider = attr.ib(
-        hash=True, default=None, validator=attr.validators.optional(attr.validators.instance_of(MasterKeyProvider))
-    )
-    source_length = attr.ib(
-        hash=True, default=None, validator=attr.validators.optional(attr.validators.instance_of(six.integer_types))
-    )
+    materials_manager = attr.ib(hash=True, default=None, validator=optional(instance_of(CryptoMaterialsManager)))
+    keyring = attr.ib(default=None, validator=optional(instance_of(Keyring)))
+    key_provider = attr.ib(hash=True, default=None, validator=optional(instance_of(MasterKeyProvider)))
+    source_length = attr.ib(hash=True, default=None, validator=optional(instance_of(six.integer_types)))
     line_length = attr.ib(
-        hash=True, default=LINE_LENGTH, validator=attr.validators.instance_of(six.integer_types)
+        hash=True, default=LINE_LENGTH, validator=instance_of(six.integer_types)
     )  # DEPRECATED: Value is no longer configurable here.  Parameter left here to avoid breaking consumers.
 
     def __attrs_post_init__(self):
         """Normalize inputs to crypto material manager."""
-        both_cmm_and_mkp_defined = self.materials_manager is not None and self.key_provider is not None
-        neither_cmm_nor_mkp_defined = self.materials_manager is None and self.key_provider is None
+        options_provided = [option is not None for option in (self.materials_manager, self.keyring, self.key_provider)]
+        provided_count = len([is_set for is_set in options_provided if is_set])
+
+        if provided_count != 1:
+            raise TypeError("Exactly one of 'materials_manager', 'keyring', or 'key_provider' must be provided")
 
-        if both_cmm_and_mkp_defined or neither_cmm_nor_mkp_defined:
-            raise TypeError("Exactly one of materials_manager or key_provider must be provided")
         if self.materials_manager is None:
-            self.materials_manager = DefaultCryptoMaterialsManager(master_key_provider=self.key_provider)
+            if self.key_provider is not None:
```

Comment:
There was an error, but now there is not.
I'm not sure I like the nested if,
What about checking `A & !B` and then `!A & B`, else `raise`?

Reply:
Now that we have three possible options, it's more complicated. That's what I'm doing in 93-97, where I throw an error if you supply anything other than exactly one of the possible options.